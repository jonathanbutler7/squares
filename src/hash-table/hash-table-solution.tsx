import { useEffect, useState } from 'react';
import {
  ColorCodes,
  getSquareNeighbors,
  Neighbors,
  getNextColorCode,
  ISquare,
  GameStatus,
  Grid,
  SelectGridSize,
  Square,
} from '../shared';

const GRID_SIZE = 4;

const getGrid = (gridSize: number) => {
  let gridValues: ISquare = {};

  for (let i = 0; i < gridSize * gridSize; i++) {
    gridValues[i] = { colorCode: ColorCodes.Red };
  }

  return gridValues;
};

const initialState = getGrid(GRID_SIZE);

const isGridAllGreen = (squares: ISquare) =>
  Object.values(squares).every(
    (square) => square.colorCode === ColorCodes.Green
  );

export const HashTableSolution = () => {
  const [squares, setSquares] = useState(initialState);
  const [clicks, setClicks] = useState(0);
  const [gridSize, setGridSize] = useState(GRID_SIZE);
  const [isGameOver, setIsGameOver] = useState(false);

  useEffect(() => {
    setIsGameOver(isGridAllGreen(squares));
  }, [squares]);

  const handleBlueClick = ({
    squareId,
    leftNeighbor,
    rightNeighbor,
    topNeighbor,
    bottomNeighbor,
  }: Neighbors & { squareId: number }) => {
    const newSquares = {
      ...(leftNeighbor && {
        [leftNeighbor]: { colorCode: ColorCodes.Blue },
      }),
      ...(rightNeighbor && {
        [rightNeighbor]: { colorCode: ColorCodes.Blue },
      }),
      ...(topNeighbor && {
        [topNeighbor]: { colorCode: ColorCodes.Blue },
      }),
      ...(bottomNeighbor && {
        [bottomNeighbor]: { colorCode: ColorCodes.Blue },
      }),
    };
    return {
      ...squares,
      [squareId]: {
        colorCode: ColorCodes.Green,
      },
      ...newSquares,
    };
  };

  const handleGreenClick = ({
    squareId,
    leftNeighbor,
    rightNeighbor,
    topNeighbor,
    bottomNeighbor,
  }: Neighbors & { squareId: number }) => {
    const newSquares = {
      ...(leftNeighbor && {
        [leftNeighbor]: {
          colorCode:
            squares[leftNeighbor].colorCode === ColorCodes.Red
              ? ColorCodes.Blue
              : ColorCodes.Green,
        },
      }),
      ...(rightNeighbor && {
        [rightNeighbor]: {
          colorCode:
            squares[rightNeighbor].colorCode === ColorCodes.Red
              ? ColorCodes.Blue
              : ColorCodes.Green,
        },
      }),
      ...(topNeighbor && {
        [topNeighbor]: {
          colorCode:
            squares[topNeighbor].colorCode === ColorCodes.Red
              ? ColorCodes.Blue
              : ColorCodes.Green,
        },
      }),
      ...(bottomNeighbor && {
        [bottomNeighbor]: {
          colorCode:
            squares[bottomNeighbor].colorCode === ColorCodes.Red
              ? ColorCodes.Blue
              : ColorCodes.Green,
        },
      }),
    };
    return {
      ...squares,
      [squareId]: {
        colorCode: getNextColorCode(squares[squareId].colorCode),
      },
      ...newSquares,
    };
  };

  const handleClick = (squareId: number) => {
    const neighbors = getSquareNeighbors({ squareId, gridSize });

    setSquares((squares) => {
      if (squares[squareId].colorCode === ColorCodes.Red) {
        return {
          ...squares,
          [squareId]: {
            colorCode: getNextColorCode(squares[squareId].colorCode),
          },
        };
      }
      if (squares[squareId].colorCode === ColorCodes.Blue) {
        return handleBlueClick({
          squareId,
          ...neighbors,
        });
      }
      if (squares[squareId].colorCode === ColorCodes.Green) {
        return handleGreenClick({
          squareId,
          ...neighbors,
        });
      } else {
        return squares;
      }
    });
  };

  const handleReset = () => {
    setClicks(0);
    setSquares(initialState);
  };

  return (
    <>
      <h1>Hash table solution </h1>
      <SelectGridSize
        options={[4, 5, 6]}
        gridSize={gridSize}
        setGridSize={setGridSize}
        handleReset={handleReset}
      />
      <Grid gridSize={GRID_SIZE}>
        {Object.values(squares).map((square, squareId) => (
          <Square
            key={squareId}
            colorCode={square.colorCode}
            onClick={() => {
              handleClick(squareId);
              setClicks((clicks) => clicks + 1);
            }}
          />
        ))}
      </Grid>
      <GameStatus
        isGameOver={isGameOver}
        clicks={clicks}
        handleReset={handleReset}
      />
    </>
  );
};
