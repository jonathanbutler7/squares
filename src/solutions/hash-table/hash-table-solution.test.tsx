import renderer from 'react-test-renderer';
import { expect, test } from 'vitest';
import { HashTableSolution } from './hash-table-solution.component';

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  return result as renderer.ReactTestRendererJSON;
}

test('Hash table solution should should render', () => {
  const component = renderer.create(<HashTableSolution />);
  let tree = toJson(component);
  expect(tree).toMatchSnapshot();
  expect(tree).toMatchSnapshot();
});
