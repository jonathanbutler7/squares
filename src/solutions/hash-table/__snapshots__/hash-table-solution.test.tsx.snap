// Vitest Snapshot v1

exports[`Hash table solution should should render 1`] = `
[
  <h1>
    Solution using hash table
  </h1>,
  <div>
    <label
      htmlFor="grid-size-options"
      style={
        {
          "marginRight": 8,
        }
      }
    >
      Change grid size:
    </label>
    <select
      name="grid-size-options"
      onChange={[Function]}
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
      <option
        value={6}
      >
        6
      </option>
    </select>
  </div>,
  <section
    style={
      {
        "border": "1px solid gray",
        "display": "grid",
        "gridTemplate": "repeat(4, 1fr) / repeat(4, 1fr)",
        "maxWidth": 400,
      }
    }
  >
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
  </section>,
  <div
    style={
      {
        "borderBottom": "1px solid gainsboro",
      }
    }
  >
    <p>
      <b>
        Clicks:
      </b>
       
      0
    </p>
    <button
      className="reset-button"
      disabled={true}
      onClick={[Function]}
      style={
        {
          "background": "#ff8fe7",
          "border": "1px solid transparent",
          "borderRadius": 4,
          "boxShadow": "none",
          "height": 40,
          "marginBottom": 16,
          "transform": "none",
          "transition": "all 250ms ease-in-out",
        }
      }
    >
      Reset game
    </button>
  </div>,
]
`;

exports[`Hash table solution should should render 2`] = `
[
  <h1>
    Solution using hash table
  </h1>,
  <div>
    <label
      htmlFor="grid-size-options"
      style={
        {
          "marginRight": 8,
        }
      }
    >
      Change grid size:
    </label>
    <select
      name="grid-size-options"
      onChange={[Function]}
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
      <option
        value={6}
      >
        6
      </option>
    </select>
  </div>,
  <section
    style={
      {
        "border": "1px solid gray",
        "display": "grid",
        "gridTemplate": "repeat(4, 1fr) / repeat(4, 1fr)",
        "maxWidth": 400,
      }
    }
  >
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
    <button
      aria-label="red square"
      className="square"
      onClick={[Function]}
      style={
        {
          "background": "rgb(226, 68, 47)",
          "border": "1px solid black",
          "height": 100,
          "width": "100%",
        }
      }
    />
  </section>,
  <div
    style={
      {
        "borderBottom": "1px solid gainsboro",
      }
    }
  >
    <p>
      <b>
        Clicks:
      </b>
       
      0
    </p>
    <button
      className="reset-button"
      disabled={true}
      onClick={[Function]}
      style={
        {
          "background": "#ff8fe7",
          "border": "1px solid transparent",
          "borderRadius": 4,
          "boxShadow": "none",
          "height": 40,
          "marginBottom": 16,
          "transform": "none",
          "transition": "all 250ms ease-in-out",
        }
      }
    >
      Reset game
    </button>
  </div>,
]
`;
